src/s_inter.c

```diff
@@ -1551,8 +1551,22 @@ void sys_bail(int n)
 
 extern void sys_exit(void);
 
+    /* before exiting Pd, closebang any canvases that are still open */
+static void sys_closebang(t_canvas *x)
+{
+    t_gobj *y;
+    for (y = x->gl_list; y; y = y->g_next)
+        if (pd_class(&y->g_pd) == canvas_class)
+            sys_closebang((t_canvas *)y);
+    canvas_closebang(x);
+}
+
 void glob_exit(void *dummy, t_float status)
 {
+    t_glist *g;
+    for (g = pd_getcanvaslist(); g; g = g->gl_next)
+        sys_closebang(g);
+
```

Comment:
Actually, we also want to *free* the root canvases. `canvas_free` calls `glist_delete` on all children, which in turn calls `canvas_loadbang` if it's a canvas - and so on recursively... This means we only miss a call to `canvas_closebang` for the root canvas itself. `sys_closebang` is not necessary.

Note that `canvas_free` already removes a root canvas from the canvas list. This means the code would be something like:
```
t_glist *c;
while (c = pd_getcanvaslist())
{
    canvas_closebang(c);
    pd_free(&c->g_pd);
}
// ...
```

NB: the order of close bangs is not "correct", but this is a general problem. I've already proposed a fix here: https://github.com/pure-data/pure-data/pull/596. With this PR, the code above would do the right thing.

Reply:
This is occurring inside of `glob_exit`, the function which shuts down Pd. For that reason, it doesn't seem like it's really necessary to free anything since everything will be freed when Pd closes.

Reply:
Just becomes the OS will eventually collect the memory, doesn't mean that you shouldn't properly release your resources. It's not only about memory, either. Objects might want to perform arbitrary tasks on shutdown and we should give them a chance to do so.

As a bonuse, the canvas destructor already sends closebangs to its children, so I do not see why we should *not* use it.